### PR TITLE
Cleanup PR

### PR DESCRIPTION
**Pull request by felixbarny**
Hey @CoderLan0668,

I could finally find some time to do a thorough review of your PR. Instead of adding comments and lots of back-and-forth, I already implemented the changes. I hope that's ok for you. If you're fine with the changes, just merge this PR and your PR will be updated.

- Reduce allocations by not wrapping callbacks
- Make plugin experimental (opt-in)
- Add @NonnullApi package-level annotations
- Use transaction.type=request, which is consistent with gRPC
- Use activate() instead of activateInScope() to reduce allocations
- Use StringBuilder.append(java.lang.CharSequence, int, int) instead of String.substring(int) to reduce allocations

### CODE REVIEW

File: apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/AlibabaMonitorFilterInstrumentation.java [hunk 1]
```diff
@@ -44,6 +46,9 @@ public AlibabaMonitorFilterInstrumentation(ElasticApmTracer tracer) {
         return named("com.alibaba.dubbo.monitor.support.MonitorFilter");
     }
 
+    /**
+     * {@link org.apache.dubbo.monitor.support.MonitorFilter#invoke(Invoker, Invocation)}
```
Comment (CoderLan0668) [inline apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/AlibabaMonitorFilterInstrumentation.java:50]: Should it be `{@link com.alibaba.dubbo.monitor.support.MonitorFilter#invoke(Invoker, Invocation)}` ?

File: apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/advice/AlibabaMonitorFilterAdvice.java [hunk 2]
```diff
@@ -57,77 +68,54 @@ public static void init(ElasticApmTracer tracer) {
...
             if (span != null) {
                 span.getTraceContext().setOutgoingTraceContextHeaders(invocation, helper);
             }
-
-            return;
+        } else {
+            // for provider side
+            transaction = tracer.startChildTransaction(invocation, helper, Invocation.class.getClassLoader());
+            if (transaction != null) {
+                transaction.activate();
```
Comment (CoderLan0668) [inline apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/advice/AlibabaMonitorFilterAdvice.java:91]: `transaction.activate()` may cause a concurrent problem, the transaction can be ended properly, and no data will be reported to apm server.

File: apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/advice/ApacheMonitorFilterAdvice.java [hunk 3]
```diff
@@ -65,58 +66,57 @@ public static void init(ElasticApmTracer tracer) {
...
-        // for provider side
-        transaction = tracer.startChildTransaction(invocation, helper, Invocation.class.getClassLoader());
-        if (transaction != null) {
-            scope = transaction.activateInScope();
-            DubboTraceHelper.fillTransaction(transaction, invocation.getInvoker().getInterface(), invocation.getMethodName());
+        } else {
+            // for provider side
+            transaction = tracer.startChildTransaction(invocation, helper, Invocation.class.getClassLoader());
+            if (transaction != null) {
+                transaction.activate();
```
Comment (CoderLan0668) [inline apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/advice/ApacheMonitorFilterAdvice.java:89]: concurrent problem as well

**Comment by CoderLan0668** [comment]
Thanks for your excellent codes, but `transaction.activate()` may cause a concurrent problem, it leads to that The Transaction can not end properly and won't be reported to the apm server, as the concurrent case is not in the unit test now.

**Comment by felixbarny** [comment]
Can you explain what that exactly means? We use activate an deactivate in all other plugins without problems. In fact, activate in scope does also call this method. The important piece is to always make sure to deactivate.

File: apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/AlibabaMonitorFilterInstrumentation.java (hunk 1 above)
Comment (felixbarny) [reply apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/AlibabaMonitorFilterInstrumentation.java:50]: Good catch!

**Comment by CoderLan0668** [comment]
Thank you！

**Pull request merged**